### PR TITLE
Homework 3 done

### PR DESCRIPTION
**Pull request by ElizavetaBB**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw3/ex1/IndexPageData.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.epam.tc.hw3.ex1;
+
+import java.util.List;
+
+public class IndexPageData {
+
+    public static String PROPERTIES_PATH = "src/test/resources/hw3/test.properties";
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/ex1/IndexPageData.java]: Upper case names uses only for final constants. Actually all data in such data classes can be final. (Valid for string, list data)

File: src/main/java/com/epam/tc/hw3/components/LoginComponent.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    private WebElement loginButton;
+
+    @FindBy(id = "user-name")
+    private WebElement username;
+
+    public LoginComponent(WebDriver driver, WebDriverWait wait) {
+        super(driver, wait);
+    }
+
+    public void login(final String login, final String password) {
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/components/LoginComponent.java]: What is the purpose of using 'final' attribute for received parameters in such methods? There and in all the same places?

File: src/main/java/com/epam/tc/hw3/fluent/DifferentElementsFluentPage.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        radioButtons.forEach(elem -> {
+            if (elem.getText().contains(name)) {
+                elem.click();
+            }
+        });
+        return this;
+    }
+
+    public WebElement getRadioButtonByName(String expected) {
+        int index = -1;
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/fluent/DifferentElementsFluentPage.java]: Why you set index = -1? Maybe it's better throw any particular exception?

File: src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    public List<String> getTextsUnderImages() {
+        return imagesTexts
+                .stream()
+                .map(WebElement::getText)
+                .collect(Collectors.toList());
+    }
+
+    public IndexFluentPage switchToFrame(WebElement frame) {
+        wait.until(ExpectedConditions.frameToBeAvailableAndSwitchToIt(frame));
+        return new IndexFluentPage(this.driver, this.wait);
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java]: Don't use approach with creating new objects to return. Try to work with the same objects not to lose any data.

File: src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    public IndexFluentPage switchToFrameWithButton() {
+        for (WebElement frame : frames) {
+            IndexFluentPage indexPage = switchToFrame(frame);
+            if (indexPage.frameButton != null) {
+                break;
+            }
+            switchToDefaultContent();
+        }
+        return new IndexFluentPage(this.driver, this.wait);
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java]: Don't use approach with creating new objects to return. Try to work with the same objects not to lose any data.

File: src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        return this;
+    }
+
+    public WebElement getFrameButton() {
+        return frameButton;
+    }
+
+    public DifferentElementsFluentPage openDifferentElementsPage() {
+        header().openDifferentElementsPage();
+        return new DifferentElementsFluentPage(this.driver, this.wait);
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java:20]: Don't use approach with creating new objects to return. Try to work with the same objects not to lose any data.

File: src/main/java/com/epam/tc/hw3/voids/DifferentElementsPage.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import com.epam.tc.hw3.components.LogComponent;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.ui.Select;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class DifferentElementsPage extends AbstractBasePage {
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/voids/DifferentElementsPage.java]: This page has bunch of the same elements with DifferentElementsFluentPage. Extend all common elements with come parent page.

File: src/main/java/com/epam/tc/hw3/voids/IndexPage.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import com.epam.tc.hw3.AbstractBasePage;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class IndexPage extends AbstractBasePage {
```
Comment (Delnote) [inline src/main/java/com/epam/tc/hw3/voids/IndexPage.java]: This page has bunch of the same elements with IndexFluentPage. Extend all common elements with come parent page.

File: src/test/java/com/epam/tc/hw3/ex2/DifferentElementsPageData.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package com.epam.tc.hw3.ex2;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DifferentElementsPageData {
+
+    public static String PROPERTIES_PATH = "src/test/resources/hw3/test.properties";
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/ex2/DifferentElementsPageData.java]: Can be moved in property file. Code duplicate from IndexPageData

File: src/test/java/com/epam/tc/hw3/ex1/IndexPageTest.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.List;
+import org.assertj.core.api.SoftAssertions;
+import org.openqa.selenium.WebElement;
+import org.testng.annotations.Test;
+
+public class IndexPageTest extends BasePageTest {
+
+    @Test
+    public void testIndexPageWithVoidPages() {
+        SoftAssertions softAssertions = new SoftAssertions();
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/ex1/IndexPageTest.java]: Please move all the same steps in both tests and classes (for DifferentElementsPageTest the same steps) in base class. Now it's code duplicate.

**Review by Delnote — COMMENTED** [review]
Be careful with code duplicate. Try to extend all common methods and steps in parent or particular util classes.

**Comment by Delnote** [comment]
I don't find jdi dependency!

File: src/main/java/com/epam/tc/hw3/fluent/DifferentElementsFluentPage.java (hunk 3 above)
Comment (ElizavetaBB) [reply src/main/java/com/epam/tc/hw3/fluent/DifferentElementsFluentPage.java]: Now the method throws NoSuchElementException

File: src/main/java/com/epam/tc/hw3/components/LoginComponent.java (hunk 2 above)
Comment (ElizavetaBB) [reply src/main/java/com/epam/tc/hw3/components/LoginComponent.java]: Deleted

File: src/test/java/com/epam/tc/hw3/ex1/IndexPageData.java (hunk 1 above)
Comment (ElizavetaBB) [reply src/test/java/com/epam/tc/hw3/ex1/IndexPageData.java]: Data in the data classes were changed to final, PROPERTY_PATH moved to PropertyReader

File: src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java (hunk 4 above)
Comment (ElizavetaBB) [reply src/main/java/com/epam/tc/hw3/fluent/IndexFluentPage.java]: Replaced with "return this" for all such cases

File: src/main/java/com/epam/tc/hw3/voids/DifferentElementsPage.java (hunk 7 above)
Comment (ElizavetaBB) [reply src/main/java/com/epam/tc/hw3/voids/DifferentElementsPage.java]: DifferentElementsFluentPage and DifferentElementsPage now have a common parent page.

File: src/main/java/com/epam/tc/hw3/voids/IndexPage.java (hunk 8 above)
Comment (ElizavetaBB) [reply src/main/java/com/epam/tc/hw3/voids/IndexPage.java]: IndexPage and IndexFluentPage now have a common parent page.

File: src/test/java/com/epam/tc/hw3/ex2/DifferentElementsPageData.java (hunk 9 above)
Comment (ElizavetaBB) [reply src/test/java/com/epam/tc/hw3/ex2/DifferentElementsPageData.java]: Property was removed

File: src/test/java/com/epam/tc/hw3/ex1/IndexPageTest.java (hunk 10 above)
Comment (ElizavetaBB) [reply src/test/java/com/epam/tc/hw3/ex1/IndexPageTest.java]: Common steps were moved to utils classes

**Comment by Delnote** [comment]
Done.

**Pull request merged**